### PR TITLE
Clean up command hierarchy around horizons

### PR DESCRIPTION
**Pull request by fknorr**
_This is the first in a series of PRs to integrate `SYNC` and `SHUTDOWN` commands into the TDAG/CDAG._

Before the introduction of task graph horizons, every `task` mapped to one or more `task_commands`, while horizons were generated on the CDAG directly. Now, CDAG horizons are simply the manifestation of TDAG horizons and therefore logically become `task_commands`, but so far this was not modelled as part of the command class hierarchy.

This PR extracts `task_command` as a common base of `execution_command` (the former `task_command`) and `horizon_command`, reducing code duplication in the process. It also removes some dead code (e.g. pseudo-critical path length tracking in `graph_generator`) that was missed when TDAG horizons were introduced.

### CODE REVIEW

File: test/test_utils.h [hunk 1]
```diff
@@ -104,15 +104,16 @@ namespace test_utils {
 
 				const detail::command_id cid = pkg.cid;
 				commands[cid] = {nid, pkg, dependencies};
-				if(pkg.cmd == detail::command_type::TASK) { by_task[std::get<detail::task_data>(pkg.data).tid].insert(cid); }
+				if(pkg.cmd == detail::command_type::EXECUTION) { by_task[std::get<detail::execution_data>(pkg.data).tid].insert(cid); }
+				if(pkg.cmd == detail::command_type::HORIZON) { by_task[std::get<detail::horizon_data>(pkg.data).tid].insert(cid); }
```
Comment (psalz) [inline test/test_utils.h:108]: And this does not break any tests? Interesting.

**Review by psalz — APPROVED** [review]
A net -38 LoC, LGTM!

File: include/command.h [hunk 2]
```diff
@@ -119,23 +108,36 @@ namespace detail {
...
 	  public:
 		task_id get_tid() const { return tid; }
 
+	  private:
+		task_id tid;
+	};
+
+	class horizon_command final : public task_command {
+		friend class command_graph;
+		using task_command::task_command;
```
Comment (PeterTh) [inline include/command.h:123]: How neat. I wasn't actively aware of this.

File: include/executor.h [hunk 3]
```diff
@@ -88,7 +88,8 @@ namespace detail {
 		template <typename Job, typename... Args>
 		void create_job(const command_pkg& pkg, const std::vector<command_id>& dependencies, Args&&... args) {
 			auto logger = execution_logger->create_context({{"job", std::to_string(pkg.cid)}});
-			if(pkg.cmd == command_type::TASK) { logger = logger->create_context({{"task", std::to_string(std::get<task_data>(pkg.data).tid)}}); }
+			if(pkg.cmd == command_type::HORIZON) { logger = logger->create_context({{"task", std::to_string(std::get<horizon_data>(pkg.data).tid)}}); }
```
Comment (PeterTh) [inline include/executor.h:91]: Maybe this could be `"horizon"`?

File: test/test_utils.h [hunk 4]
```diff
@@ -177,15 +179,19 @@ namespace test_utils {
...
 			if(most_recently_generated_task_horizon != most_recently_built_task_horizon) {
 				most_recently_built_task_horizon = most_recently_generated_task_horizon;
 				if(most_recently_built_task_horizon != nullptr) {
-					auto htid = most_recently_built_task_horizon->get_id();
-					get_graph_generator().build_task(htid, {nullptr});
+					const auto htid = most_recently_built_task_horizon->get_id();
+					// naive_split does not really do anything for horizons, but this mirrors the behavior of scheduler::schedule exactly.
+					detail::naive_split_transformer naive_split(num_nodes, num_nodes);
+					get_graph_generator().build_task(htid, {&naive_split});
+					return htid;
```
Comment (PeterTh) [inline test/test_utils.h:191]: As I mentioned during the meeting, I think having to reproduce a lot of the logic in the test utils is not ideal. But I also don't have a better idea for unit tests which shouldn't have the entire runtime in the background.

**Review by PeterTh — APPROVED** [review]

File: include/command.h (hunk 2 above)
Comment (fknorr) [reply include/command.h:123]: Re-exporting base-class constructors? Yeah, that is pretty neat!

File: test/test_utils.h (hunk 1 above)
Comment (fknorr) [reply test/test_utils.h:108]: `by_task` is only used when passing a `task_id` to `cdag_inpector::get_commands()`, which was never done for horizon tasks as far as I'm aware.

File: test/test_utils.h (hunk 4 above)
Comment (fknorr) [reply test/test_utils.h:191]: I think we could reasonably manage the set of active transformers in `graph_generator` and get rid of the `native_split` construction (and vector allocation) here. Thoughts @psalz ?

**Review by facuMH — APPROVED** [review]

File: test/test_utils.h (hunk 4 above)
Comment (psalz) [reply test/test_utils.h:191]: Hmm, I'm not sure. While the scheduler is pretty bare-bones at the moment, I have at various occasions experimented with multiple/different transformers depending on task types, hints etc -- which is something that I don't think the graph generator should be aware of. We can certainly revisit the current scheduler/transformer/ggen design, but I'd say its out of scope for this PR.

Comment (fknorr) [reply test/test_utils.h:191]: Let's just keep it in mind then.

**Pull request merged**